### PR TITLE
Filedownloads url - Adhere to w3c standards

### LINKED ISSUE #11458: [🚀 Feature]: List all downloaded files in a session (Grid)
### Feature and motivation

#11443 is a proposal to implement file download on the client bindings. However, to download a file you need to know the file name, and this might be tricky given that most of the time the application under test does not let you choose the name that will be used when the file is downloaded.

We should have an endpoint that lists the downloaded files in the configured download directory in Grid. Then the user can choose the name of the one they want to download.

### Usage example

Something along the lines of:

`/session/:sessionId:/se/file`

Which could return:

```json
[
	{
		"name":"input.txt",
		"timestamp": "<timestamp with creation date"
	},
	{
		"name":"anotherFile.txt",
		"timestamp": "<timestamp with creation date"
	}
]
```

But I am not 100% that is the W3C WebDriver response payload structure, we should adapt to that one. @titusfortner do you know?

The main challenge in this issue is to decide if we need to list files that belong to a session or if it is OK to list all downloaded files.

### LINKED ISSUE #11466: [🐛 Bug]: Grid File download capability should be w3c compliant
### What happened?

I'm sorry I didn't weigh in on #11277 before it was merged. I'm thinking of this specifically for how to get this properly consumed by the bindings as per #11443 

2 things to bring the grid code more in line with other endpoints.
1. Use POST if there is info to send to an endpoint instead of query string
2. Return payload needs to be nested inside a `value` key.

Current implementation of file endpoint is:
```
POST /session/:sessionId:/se/file
file: <zippedFile>
```

To minimize confusion, I think the new feature should use `files` instead of `file`:
```
POST /session/:sessionId:/se/files
{fileName: <filename>}
```
and for #11458 
```
GET /session/:sessionId:/se/files
```

### How can we reproduce the issue?

```shell
n/a
```

### Relevant log output

```shell
n/a
```

### Operating System

n/a

### Selenium version

trunk

### What are the browser(s) and version(s) where you see this issue?

n/a

### What are the browser driver(s) and version(s) where you see this issue?

n/a

### Are you using Selenium Grid?

trunk

### PR DESCRIPTION
**Pull request by krmahadevan**
Fixes: #11466 #11458

New format

`POST /se/files`

Request:

`{ "filename": "file" }`

Response:
```
{
	"value": {
	"filename": "filename",
	"contents": "ContentsGoHere"
	}
}
```

`GET /se/files`

Request:
NONE

Response:
```
{
	"value": {
		"files":[
		"1",
		"2",
		"3"
		]
	}
}
```

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@diemol @titusfortner 

This PR takes care of ONLY streamlining the url to be compliant with `w3c` urls and also adds the capability to list files for download.

The PR for https://github.com/SeleniumHQ/selenium/issues/11443 will ensure that the downloads path references are all removed.

I thought it would be good if I go at this requirement on a piece by piece

**Comment by surajverma255** [comment]
Hi Community,
Have just started to explore the Open Source Community.
As a Noobie, I feel we should have a quick description of the changes someone is pushing which will make it easier to understand the context of change :)

**Comment by titusfortner** [comment]
@krmahadevan I agree with going piece by piece, so I broke up the issue above into 5 components, this being two I think. Will review soon.

@surajverma255 welcome! Please read the issues linked in the PR to better understand the context.

**Comment by titusfortner** [comment]
yeah, ok, @diemol will have to review the code as I'm lost in this part of Selenium, but the syntax looks correct to me.

Diego previously proposed to add file timestamp info with the GET response. To be honest, it would be a lot easier not to include it. Especially if we are already breaking things out by session, but we should agree on that bit.

File: java/src/org/openqa/selenium/grid/node/local/LocalNode.java [hunk 1]
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
     if (!dir.isDirectory()) {
       throw new WebDriverException(String.format("Invalid directory: %s.", downloadsPath));
     }
-    String filename = req.getQueryParameter("filename");
+    if (req.getMethod().equals(HttpMethod.GET)) {
+      //User wants to list files that can be downloaded
+      List<String> collected = Arrays.stream(
+        Optional.ofNullable(dir.listFiles()).orElse(new File[]{})
+      ).map(File::getName).collect(Collectors.toList());
+      ImmutableMap<String, Object> data = ImmutableMap.of("filenames", collected);
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/node/local/LocalNode.java]:
```suggestion
      ImmutableMap<String, Object> data = ImmutableMap.of("names", collected);
```

File: java/src/org/openqa/selenium/grid/node/local/LocalNode.java [hunk 2]
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
...
+      ImmutableMap<String, Map<String, Object>> result = ImmutableMap.of("value",data);
+      return new HttpResponse().setContent(asJson(result));
+    }
+
+    String raw = string(req);
+    if (raw.isEmpty()) {
+      throw new WebDriverException(
+        "Please specify file to download in payload as {\"filename\": \"fileToDownload\"}");
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/node/local/LocalNode.java]:
```suggestion
        "Please specify file to download in payload as {\"name\": \"fileToDownload\"}");
```

File: java/src/org/openqa/selenium/grid/node/local/LocalNode.java [hunk 3]
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
...
+    }
+
+    Map<String, Object> incoming = JSON.toType(raw, Json.MAP_TYPE);
+    String filename = Optional.ofNullable(incoming.get("filename"))
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/node/local/LocalNode.java]:
```suggestion
    String filename = Optional.ofNullable(incoming.get("name"))
```

File: java/src/org/openqa/selenium/grid/node/local/LocalNode.java [hunk 4]
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
...
+      .map(Object::toString)
+      .orElseThrow(
+        () -> new WebDriverException(
+          "Please specify file to download in payload as {\"filename\": \"fileToDownload\"}"));
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/node/local/LocalNode.java]:
```suggestion
          "Please specify file to download in payload as {\"name\": \"fileToDownload\"}"));
```

**Review by diemol — APPROVED** [review]
Just very minor comments, as in the endpoint already shows it is about files. 

Thank you, @krmahadevan!

**Comment by krmahadevan** [comment]
> Just very minor comments, as in the endpoint already shows it is about files.
> 
> Thank you, @krmahadevan!

Thank you @diemol. i have fixed the review comments as an additional commit and also rebased off of `trunk`. Can I go ahead and merge this or would you or @titusfortner do that? First PR after a committer :) So I dont know yet if I can hit the `merge` button (Even though I can see it.. :D )

**Comment by diemol** [comment]
@krmahadevan sure, go for it.

**Pull request merged**